src/hpc/boehm_gc.h

```diff
@@ -0,0 +1,647 @@
+
+#ifdef  COUNT_BAGS
+	/* update the statistics                                               */
+	InfoBags[type].sizeLive += new_size - old_size;
+	InfoBags[type].sizeAll  += new_size - old_size;
+#endif
+	SizeAllBags             += new_size - old_size;
+
+#ifndef DISABLE_GC
+	alloc_size = GC_size(PTR_BAG(bag)-HEADER_SIZE);
+	/* An alternative implementation would be to compare
+	 * new_size <= alloc_size in the following test in order
+	 * to avoid reallocations for alternating contractions
+	 * and expansions. However, typed allocation in the Boehm
+	 * GC stores layout information in the last word of a memory
+	 * block and we may accidentally overwrite this information,
+	 * because GC_size() includes that extraneous word when
+	 * returning the size of a memory block.
+	 *
+	 * This is technically a bug in GC_size(), but until and
+	 * unless there is an upstream fix, we'll do it the safe
+	 * way.
+	 */
+	if ( new_size <= old_size
+			 && HEADER_SIZE*sizeof(Bag) + new_size >= alloc_size * 3/4) {
+#else
+	if (new_size <= old_size) {
+#endif /* DISABLE_GC */
+
+			/* change the size word                                            */
+#ifdef USE_NEWSHAPE
+		*(*bag-2) = (new_size << 16 | type);
+#else
+		*(*bag-2) = new_size;
+#endif
+	}
+
+
+	/* if the bag is enlarged                                              */
+	else {
+			alloc_size = HEADER_SIZE*sizeof(Bag) + new_size;
+			if (new_size == 0)
+					alloc_size++;
+#ifndef DISABLE_GC
+			dst = AllocateBagMemory(TabMarkTypeBags[type], type, alloc_size);
+#else
+			dst       = malloc( alloc_size );
+			memset(dst, 0, alloc_size);
+#endif
+
+			/* leave magic size-type word  for the sweeper, type must be 255   */
+#ifdef USE_NEWSHAPE
+			*dst++ = (Bag)(new_size << 16 | type);
+#else
+			/* enter the new size-type word                                    */
+			*dst++ = (Bag)type;
+			*dst++ = (Bag)new_size;
+#endif
+
+
+			*dst++ = bag;
+			/* set the masterpointer                                           */
+			src = PTR_BAG(bag);
+			PTR_BAG(bag) = dst;
+
+			if (dst != src) {
+					memcpy( dst, src, old_size < new_size ? old_size : new_size );
+			} else if (new_size < old_size) {
+					memset(dst+new_size, 0, old_size - new_size);
+			}
+	}
+	/* return success                                                      */
+	return 1;
+}
+
+
+/*****************************************************************************
+** The following functions are not required by boehm, so empty implementations
+** are provided
+**
+*/
+
+void            InitMsgsFuncBags (
+    TNumMsgsFuncBags    msgs_func )
+{ }
+
+void InitGlobalBag (
+    Bag *               addr,
+    const Char *        cookie )
+{ }
+
+void SortGlobals( UInt byWhat )
+{ }
+
+Bag * GlobalByCookie(
+       const Char * cookie )
+{
+    return (Bag *) 0;
+}
+
+void CallbackForAllBags(
+     void (*func)(Bag) )
+{ }
+
+void StartRestoringBags( UInt nBags, UInt maxSize)
+{ }
+
+Bag NextBagRestoring( UInt size, UInt type)
+{ return 0; }
```

Comment:
Also, the empty function bodies above are weird, but not as weird as that one :)